File: Templates/objective-c/ProximityContent/ProximityContent/Estimote/BeaconContentCache.m
//
//  BeaconContentCache.m
//  ProximityContent
//

#import "BeaconContentCache.h"

@interface BeaconContentCache ()

@property (nonatomic, readonly) id<BeaconContentFactory> beaconContentFactory;

@property (nonatomic) NSMutableDictionary *cachedContent;

@end

@implementation BeaconContentCache

- (instancetype)initWithBeaconContentFactory:(id<BeaconContentFactory>)beaconContentFactory {
    self = [super init];
    if (self) {
        _beaconContentFactory = beaconContentFactory;
        _cachedContent = [NSMutableDictionary new];
    }
    return self;
}

- (void)contentForBeaconID:(BeaconID *)beaconID completion:(void (^)(id))completion {
    id content = [self.cachedContent objectForKey:beaconID];
    if (content) {
        completion(content);
    } else {
        [self.beaconContentFactory contentForBeaconID:beaconID completion:^(id content) {
            self.cachedContent[beaconID] = content;
            completion(content);
        }];
    }
}

@end

File: Templates/swift/ProximityContent/ProximityContent/Estimote/BeaconContentCache.swift
//
//  BeaconContentCache.swift
//  ProximityContent
//

class BeaconContentCache {

    private let beaconContentFactory: BeaconContentFactory

    private var cachedContent = Dictionary<BeaconID, AnyObject>()

    init(beaconContentFactory: BeaconContentFactory) {
        self.beaconContentFactory = beaconContentFactory
    }

    func contentForBeaconID(beaconID: BeaconID, completion: (AnyObject) -> ()) {
        if let content: AnyObject = self.cachedContent[beaconID] {
            completion(content)
        } else {
            self.beaconContentFactory.contentForBeaconID(beaconID) { (content) in
                self.cachedContent[beaconID] = content
                completion(content)
            }
        }
    }

}
